File: bindings/typescript/packages/aster/src/interceptors/capability.ts
/**
 * Capability interceptor — validates method-level access control.
 *
 * Checks that the caller's role (from admission attributes) satisfies
 * the method's capability requirement.
 */

import type { Interceptor } from './base.js';
import { CallContext } from './base.js';
import { RpcError, StatusCode } from '../status.js';
import type { CapabilityRequirement } from '../service.js';

/**
 * Normalise a `requires` value to a `CapabilityRequirement` object.
 *
 * `@Rpc({ requires: Role.ADMIN })` passes a bare string for the common
 * single-role case; `anyOf()`/`allOf()` produce a structured object. Both
 * forms are accepted here so the interceptor logic stays uniform.
 */
function normaliseRequirement(req: unknown): CapabilityRequirement | null {
  if (req == null) return null;
  if (typeof req === 'string') {
    return { kind: 'role', roles: [req] };
  }
  if (typeof req === 'object' && req !== null && 'kind' in req && 'roles' in req) {
    return req as CapabilityRequirement;
  }
  return null;
}

export class CapabilityInterceptor implements Interceptor {
  private requirements = new Map<string, CapabilityRequirement>();

  /** Register a capability requirement for a method. */
  setRequirement(service: string, method: string, req: CapabilityRequirement | string): void {
    const normalised = normaliseRequirement(req);
    if (normalised) {
      this.requirements.set(`${service}/${method}`, normalised);
    }
  }

  async onRequest(ctx: CallContext, request: unknown): Promise<unknown> {
    const key = `${ctx.service}/${ctx.method}`;
    const req = this.requirements.get(key);
    if (!req) return request;

    // aster.role is a comma-separated list: "ops.status,ops.logs,ops.admin"
    const roleStr = ctx.attributes['aster.role'] ?? '';
    const callerRoles = new Set(roleStr.split(',').map(r => r.trim()).filter(Boolean));

    switch (req.kind) {
      case 'role':
        if (!callerRoles.has(req.roles[0]!)) {
          throw new RpcError(StatusCode.PERMISSION_DENIED, `requires role: ${req.roles.join(', ')}`);
        }
        break;
      case 'any_of':
        if (!req.roles.some(r => callerRoles.has(r))) {
          throw new RpcError(StatusCode.PERMISSION_DENIED, `requires any of: ${req.roles.join(', ')}`);
        }
        break;
      case 'all_of':
        if (!req.roles.every(r => callerRoles.has(r))) {
          throw new RpcError(StatusCode.PERMISSION_DENIED, `requires all of: ${req.roles.join(', ')}`);
        }
        break;
    }
    return request;
  }
}

File: bindings/python/aster/interceptors/capability.py
"""Capability interceptor -- Gate 3 method-level access control.

Evaluates ``CapabilityRequirement`` from the service and method definitions
against the caller's admission attributes (``CallContext.attributes``).

Both the service-level ``requires`` (baseline) and the method-level
``requires`` must be satisfied (conjunction).  If neither is defined the
call passes through unconditionally.
"""

from __future__ import annotations

import logging
from typing import Any

from aster.contract.identity import CapabilityRequirement
from aster.interceptors.base import CallContext, Interceptor
from aster.service import ServiceInfo
from aster.status import RpcError, StatusCode
from aster.trust.rcan import evaluate_capability

logger = logging.getLogger(__name__)


def _normalize_requirement(req: Any) -> CapabilityRequirement | None:
    """Normalize a requires= value to a CapabilityRequirement.

    Accepts:
    - None -> None
    - CapabilityRequirement -> pass through
    - str or str enum -> wrap as ROLE requirement
    """
    if req is None:
        return None
    if isinstance(req, CapabilityRequirement):
        return req
    if isinstance(req, str):
        from aster.contract.identity import CapabilityKind
        return CapabilityRequirement(kind=CapabilityKind.ROLE, roles=[str(req)])
    return req


class CapabilityInterceptor(Interceptor):
    """Enforces capability requirements on incoming RPC calls.

    Constructed with a mapping of service names to their ``ServiceInfo`` so
    that both service-level and method-level ``requires`` can be looked up
    at request time.
    """

    def __init__(self, service_map: dict[str, ServiceInfo]) -> None:
        self._service_map = service_map

    async def on_request(self, ctx: CallContext, request: object) -> object:
        svc_info = self._service_map.get(ctx.service)
        if svc_info is None:
            # Service not known to this interceptor -- pass through.
            return request

        # Service-level baseline requirement.
        svc_req = _normalize_requirement(svc_info.requires)
        # Method-level requirement.
        method_req = None
        method_info = svc_info.get_method(ctx.method)
        if method_info is not None:
            method_req = _normalize_requirement(method_info.requires)

        # No requirements at all -- pass through.
        if svc_req is None and method_req is None:
            return request

        # Evaluate service-level requirement.
        if svc_req is not None:
            if not evaluate_capability(svc_req, ctx.attributes):
                logger.warning(
                    "Capability denied: service=%s method=%s peer=%s "
                    "(service-level requirement not met)",
                    ctx.service, ctx.method, ctx.peer,
                )
                raise RpcError(
                    StatusCode.PERMISSION_DENIED,
                    f"capability check failed for service '{ctx.service}'",
                )

        # Evaluate method-level requirement.
        if method_req is not None:
            if not evaluate_capability(method_req, ctx.attributes):
                logger.warning(
                    "Capability denied: service=%s method=%s peer=%s "
                    "(method-level requirement not met)",
                    ctx.service, ctx.method, ctx.peer,
                )
                raise RpcError(
                    StatusCode.PERMISSION_DENIED,
                    f"capability check failed for method '{ctx.service}.{ctx.method}'",
                )

        return request
